### PR TITLE
Fixed creationContext not being passed to abstract factory canCreate()

### PR DESCRIPTION
**Pull request by kynx**
Fix for #78

### CODE REVIEW

**Comment by kynx** [comment]
Ping @weierophinney @ezimuel

**Comment by Ocramius** [comment]
@kynx merged, thanks!

File: test/AbstractPluginManagerTest.php [hunk 1]
```diff
@@ -349,4 +350,17 @@ public function testPassingServiceInstanceViaConfigureShouldRaiseExceptionForInv
             stdClass::class => new stdClass(),
         ]]);
     }
+
+    public function testAbstractFactoryGetsCreationContext()
+    {
+        $serviceManager = new ServiceManager();
+        $pluginManager = new TestAsset\SimplePluginManager($serviceManager);
+        $abstractFactory = $this->prophesize(AbstractFactoryInterface::class);
+        $abstractFactory->canCreate($serviceManager, 'foo')
```
Comment (bakura10) [inline test/AbstractPluginManagerTest.php]: shouldBeCalled() should be added (as far as I remember, without shouldBeCalled this assertion does nothing actually, and changing foo to something completely different does not throw any error... but I may be wrong. Mind testing @Ocramius ?

File: test/AbstractPluginManagerTest.php [hunk 2]
```diff
@@ -349,4 +350,17 @@ public function testPassingServiceInstanceViaConfigureShouldRaiseExceptionForInv
...
+            ->willReturn(true);
+        $abstractFactory->__invoke($serviceManager, 'foo', null)
+            ->willReturn(new InvokableObject());
```
Comment (bakura10) [inline test/AbstractPluginManagerTest.php]: Same, ->shouldBeCalled() is necessary.

Comment (Ocramius) [inline test/AbstractPluginManagerTest.php]: `shouldBeCalled` is not necessary, as the purpose of the test is not to verify that the factory was called, but that the `InvokableObject` was returned (and the factory received those parameters in between).

`shouldBeCalled` is `self::atLeastOnce()`, IIRC.

Comment (Ocramius) [inline test/AbstractPluginManagerTest.php]: Also: every time I read prophecy code I cringe >.<

File: test/AbstractPluginManagerTest.php (hunk 1 above)
Comment (kynx) [inline test/AbstractPluginManagerTest.php]: True. But the intention was to catch the situation where `canCreate()` was called with the `$pluginManager` instead of the `$serviceManager`, which it does (in a rather ugly fashion ;)

File: test/AbstractPluginManagerTest.php (hunk 2 above)
Comment (bakura10) [inline test/AbstractPluginManagerTest.php]: Haha @danizord absolutely wanted we used Prophecy and I actually quite like it now :). I've tested, you're right. I don't know what I thought that :/.

Comment (danizord) [inline test/AbstractPluginManagerTest.php]: `shouldBeCalled()` is not mandatory here, but it is useful for debugging. If the method does not get called, then I'd prefer the test to fail early telling that it was not called instead instad of failing on `assertInstanceOf()` where I can't see why it failed.

Comment (Ocramius) [inline test/AbstractPluginManagerTest.php]: Send another patch.

@danizord stop brain-washing @bakura10 :-P

Dynamic methods are bad, mmkay? :-P

Comment (bakura10) [inline test/AbstractPluginManagerTest.php]: If only this young dude would brain-wash me only on this... :p

Comment (danizord) [inline test/AbstractPluginManagerTest.php]: >Dynamic methods are bad, mmkay? :-P

\*cough\*, I've learned from GandalPHP that magic is good when it works :)

Comment (Ocramius) [inline test/AbstractPluginManagerTest.php]: @danizord that's why I wrote a 2 Mb library to avoid having magic methods ;-)

Comment (danizord) [inline test/AbstractPluginManagerTest.php]: I see, but they don't hurt in this context (tests) IMHO. It just makes the test a lot more readable than:

```php
$this->getMock(MyStuff::class, [], [], '', false)->expects($this->once())->method('doSomething')->with($this->equalTo('something'));
```

Comment (Ocramius) [inline test/AbstractPluginManagerTest.php]: Yeah, I like (and read) that syntax better, heh.

Comment (danizord) [inline test/AbstractPluginManagerTest.php]: Rly? Damn, that was my unique argument. :joy:
